File: app/src/main/java/com/example/ms4/MainActivity.java
package com.example.ms4;

import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import java.util.HashMap;

public class MainActivity extends AppCompatActivity {
    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        //1.类的加载过程
        //2.hashmap
        //3.锁的区别底层实现
        //4.okhttp 比原来的优势，拦截器，为什么要有拦截器
        //5.glide的优点
    }
    //类的加载过程
    private void question1(){
        //a.在加载class文件的时候，JVM会先加载类中的所有静态成员( 方法，变量，静态代码块 )都加载到方法区class文件的所处静态区中
        //b.当把所有的静态成员加载完成之后，开始给类中的所有静态成员变量进行默认初始化
        //c.当类中的所有静态成员变量默认初始化之后，接着开始给所有静态成员变量显示赋值。
        //d.当类中所有的静态成员变量显示赋值结束之后，静态代码块才会运行。
        //e.当静态代码块执行结束之后，才表示class文件加载完成
    }
    //hashmap
    private void question2(){
        //数组：存储区间连续，占用内存严重，寻址容易，插入删除困难；
        //链表：存储区间离散，占用内存比较宽松，寻址困难，插入删除容易；
        //Hashmap综合应用了这两种数据结构，实现了寻址容易，插入删除也容易。
        //HashMap 内部结构：可以看作是数组和链表结合组成的复合结构，数组被分为一个个桶（bucket），每个桶存储有
          //一个或多个Entry对象，每个Entry对象包含三部分key（键）、value（值），next(指向下一个Entry），通
          //过哈希值决定了Entry对象在这个数组的寻址
    }
}

File: app/src/main/java/com/example/kotlin/kotlin_funandlambda/MainActivity.kt
package com.example.kotlin_funandlambda

import android.os.Bundle
import android.util.Log
import androidx.appcompat.app.AppCompatActivity

/**
 * 函数
 */
class MainActivity :AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        //函数调用
        val double = Double(2)
        //具名函数调用
        reformat("111",word = '1')
        asList(1,2,3)
        Sample().foo()
    }
    //1.函数声明
    /**
     * 函数声明
     * 使用fun 声明函数
     */
    fun Double(x:Int):Int{
        return 2*x
    }

    //2.参数
    /**
     * 参数
     * name:type 参数用逗号分开，每个参数必须有显示类型
     */
    fun powerOff(number:Int,total:Int){

    }

    /**
     * 3.默认参数
     * 函数参数可以有默认值，当省略相应的参数时使用默认值。可以减少重载数量
     */
    fun read(b:Array<Byte>,offet:Int,len:Int = b.size){

    }

    /**
     * 4.覆盖带有默认参数的方法
     * 必须从签名中省略默认值
     */
     open class A{
        open fun foo(age:Int = 10){

        }
    }
    class B:A(){
        //不能有默认值
        override fun foo(age: Int) {
            super.foo(age)
        }
    }
    /**
     * 5.如果一个默认参数在一个无默认参数之前，那么该默认值只能通过具名参数调用该函数来使用
     */
    fun foo2(bar:Int = 0,baz:Int){

    }

    /**
     * 6.具名参数
     * 在调用函数时，使用具名的函数当参数
     */
    fun reformat(str:String,normal:Boolean = true,upper:Boolean = false,word:Char = ' '){
       Log.d("reformat","str:"+str+"word:"+word)
    }

    /**
     * 7.返回Unit
     * 函数无返回值
     * :Unit 可以省略
     */
    fun printName(name:String):Unit{
        Log.d("printName","name:"+name)
    }

    /**
     * 8.单表达式函数
     * 当函数返回单个表达式时，可以省略花括号，在等于号后加上表达体
     */
    fun double(x:Int):Int= x*2
    fun double2(x:Int) = x*2 //返回类型由编辑器推断出来，函数返回类型可以省略

    /**
     * 9.可变数量的参数Varargs
     */
    fun <T> asList(vararg t:T ):List<T>{
        val result = ArrayList<T>()
        for(ts in t){
            result.add(ts)
        }
        return result
    }

    /**
     * 10.成员函数
     */
    class Sample(){
      fun  foo(){
          Log.d("Sample","sample:foo()")
      }
    }


    /**
     * 泛型函数
     */
    fun <T> getData(item:T):List<T>{
        val list = arrayListOf<T>()
        list.add(item)
        return list
    }




}